### PR TITLE
Sema: Fix existential member access type erasure around metatypes

### LINKED ISSUE #72941: Compiler crash with dynamic generic
### Description

The compiler crashed for a complex generic function case.

### Reproduction

Try and compile the following Swift file:
```swift
import Foundation


protocol MyProto : Codable, Sendable {}
struct MyProtoImplem : MyProto {}
func decodeStuff<T : Codable & Sendable>(_ type: T.Type = T.self) throws -> T {
   throw Err()
}

let t: MyProto.Type = MyProtoImplem.self
if let r = try? decodeStuff(t.optionalSelf) {
   print("ok: \(r)")
}


extension MyProto {
   static var optionalSelf: Optional<Self>.Type {
      return Optional<Self>.self
   }
}

struct Err : Error {}
```

### Stack dump

```text
Stack dump:
0.	Program arguments: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/swift-frontend -frontend -interpret toto.swift -Xllvm -aarch64-use-tbi -enable-objc-interop -stack-check -sdk /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.4.sdk -color-diagnostics -new-driver-path /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/swift-driver -empty-abi-descriptor -resource-dir /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/swift -module-name toto -disable-clang-spi -target-sdk-version 14.4 -target-sdk-name macosx14.4 -external-plugin-path /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.4.sdk/usr/lib/swift/host/plugins#/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.4.sdk/usr/bin/swift-plugin-server -external-plugin-path /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.4.sdk/usr/local/lib/swift/host/plugins#/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.4.sdk/usr/bin/swift-plugin-server -external-plugin-path /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/usr/lib/swift/host/plugins#/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/usr/bin/swift-plugin-server -external-plugin-path /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/usr/local/lib/swift/host/plugins#/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/usr/bin/swift-plugin-server -plugin-path /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/swift/host/plugins -plugin-path /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/local/lib/swift/host/plugins
1.	Apple Swift version 5.10 (swiftlang-5.10.0.13 clang-1500.3.9.4)
2.	Compiling with the current language version
3.	While evaluating request TypeCheckSourceFileRequest(source_file "toto.swift")
4.	While type-checking statement at [toto.swift:10:1 - line:12:1] RangeText="if let r = try? decodeStuff(t.optionalSelf) {
	print("ok: \(r)")
"
5.	While type-checking statement at [toto.swift:10:1 - line:12:1] RangeText="if let r = try? decodeStuff(t.optionalSelf) {
	print("ok: \(r)")
"
6.	While type-checking expression at [toto.swift:10:12 - line:10:43] RangeText="try? decodeStuff(t.optionalSelf"
7.	While type-checking-target starting at toto.swift:10:17
8.	While evaluating request AbstractGenericSignatureRequest(NULL, {τ_0_0}, {τ_0_0 : Optional<MyProto>})
Stack dump without symbol names (ensure you have llvm-symbolizer in your PATH or set the environment var `LLVM_SYMBOLIZER_PATH` to point to it):
0  swift-frontend           0x000000010826bf3c llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) + 56
1  swift-frontend           0x000000010826b0f8 llvm::sys::RunSignalHandlers() + 112
2  swift-frontend           0x000000010826c544 SignalHandler(int) + 360
3  libsystem_platform.dylib 0x000000019be97584 _sigtramp + 56
4  swift-frontend           0x000000010454c228 swift::AbstractGenericSignatureRequest::evaluate(swift::Evaluator&, swift::GenericSignatureImpl const*, llvm::SmallVector<swift::GenericTypeParamType*, 2u>, llvm::SmallVector<swift::Requirement, 2u>) const + 2712
5  swift-frontend           0x000000010454c228 swift::AbstractGenericSignatureRequest::evaluate(swift::Evaluator&, swift::GenericSignatureImpl const*, llvm::SmallVector<swift::GenericTypeParamType*, 2u>, llvm::SmallVector<swift::Requirement, 2u>) const + 2712
6  swift-frontend           0x0000000103f92c94 swift::SimpleRequest<swift::AbstractGenericSignatureRequest, llvm::PointerIntPair<swift::GenericSignature, 3u, swift::OptionSet<swift::GenericSignatureErrorFlags, unsigned int>, llvm::PointerLikeTypeTraits<swift::GenericSignature>, llvm::PointerIntPairInfo<swift::GenericSignature, 3u, llvm::PointerLikeTypeTraits<swift::GenericSignature>>> (swift::GenericSignatureImpl const*, llvm::SmallVector<swift::GenericTypeParamType*, 2u>, llvm::SmallVector<swift::Requirement, 2u>), (swift::RequestFlags)2>::evaluateRequest(swift::AbstractGenericSignatureRequest const&, swift::Evaluator&) + 284
7  swift-frontend           0x000000010446984c llvm::Expected<swift::AbstractGenericSignatureRequest::OutputType> swift::Evaluator::getResultUncached<swift::AbstractGenericSignatureRequest>(swift::AbstractGenericSignatureRequest const&) + 888
8  swift-frontend           0x0000000104469348 llvm::Expected<swift::AbstractGenericSignatureRequest::OutputType> swift::Evaluator::getResultCached<swift::AbstractGenericSignatureRequest, (void*)0>(swift::AbstractGenericSignatureRequest const&) + 676
9  swift-frontend           0x0000000104467f98 swift::buildGenericSignature(swift::ASTContext&, swift::GenericSignature, llvm::SmallVector<swift::GenericTypeParamType*, 2u>, llvm::SmallVector<swift::Requirement, 2u>) + 260
10 swift-frontend           0x00000001041e06b4 swift::ASTContext::getOpenedExistentialSignature(swift::Type, swift::GenericSignature) + 732
11 swift-frontend           0x00000001041df1b8 swift::GenericEnvironment::forOpenedExistential(swift::Type, swift::GenericSignature, swift::UUID) + 188
12 swift-frontend           0x00000001041deeec swift::OpenedArchetypeType::get(swift::CanType, swift::GenericSignature, llvm::Optional<swift::UUID>) + 272
13 swift-frontend           0x00000001045988bc swift::TypeBase::openAnyExistentialType(swift::OpenedArchetypeType*&, swift::GenericSignature) + 188
14 swift-frontend           0x0000000103d4c2c8 swift::constraints::ConstraintSystem::openExistentialType(swift::Type, swift::constraints::ConstraintLocator*) + 132
15 swift-frontend           0x0000000103c7cfe0 matchCallArguments(swift::constraints::ConstraintSystem&, swift::FunctionType*, swift::ArgumentList*, llvm::ArrayRef<swift::AnyFunctionType::Param>, llvm::ArrayRef<swift::AnyFunctionType::Param>, swift::constraints::ConstraintKind, swift::constraints::ConstraintLocatorBuilder, llvm::Optional<swift::constraints::TrailingClosureMatching>, llvm::SmallVectorImpl<std::__1::pair<swift::TypeVariableType*, swift::OpenedArchetypeType*>>&) + 9132
16 swift-frontend           0x0000000103c79e00 swift::constraints::ConstraintSystem::simplifyApplicableFnConstraint(swift::Type, swift::Type, llvm::Optional<swift::constraints::TrailingClosureMatching>, swift::OptionSet<swift::constraints::ConstraintSystem::TypeMatchFlags, unsigned int>, swift::constraints::ConstraintLocatorBuilder) + 4452
17 swift-frontend           0x0000000103c84958 swift::constraints::ConstraintSystem::addConstraintImpl(swift::constraints::ConstraintKind, swift::Type, swift::Type, swift::constraints::ConstraintLocatorBuilder, bool) + 1612
18 swift-frontend           0x0000000103c51e48 swift::constraints::ConstraintSystem::addConstraint(swift::constraints::ConstraintKind, swift::Type, swift::Type, swift::constraints::ConstraintLocatorBuilder, bool) + 80
19 swift-frontend           0x0000000103c37c04 (anonymous namespace)::ConstraintGenerator::visitApplyExpr(swift::ApplyExpr*) + 728
20 swift-frontend           0x0000000103c26204 (anonymous namespace)::ConstraintWalker::walkToExprPost(swift::Expr*) + 740
21 swift-frontend           0x00000001042da670 (anonymous namespace)::Traversal::visitOptionalTryExpr(swift::OptionalTryExpr*) + 116
22 swift-frontend           0x00000001042d93e0 (anonymous namespace)::Traversal::visit(swift::Expr*) + 268
23 swift-frontend           0x0000000103c20770 swift::constraints::ConstraintSystem::generateConstraints(swift::Expr*, swift::DeclContext*, bool) + 220
24 swift-frontend           0x0000000103c2007c swift::constraints::ConstraintSystem::generateConstraints(swift::constraints::SyntacticElementTarget&, swift::FreeTypeVariableBinding) + 1260
25 swift-frontend           0x0000000103c9f750 swift::constraints::ConstraintSystem::solve(swift::constraints::SyntacticElementTarget&, swift::FreeTypeVariableBinding) + 932
26 swift-frontend           0x0000000103e8bb48 swift::TypeChecker::typeCheckTarget(swift::constraints::SyntacticElementTarget&, swift::OptionSet<swift::TypeCheckExprFlags, unsigned int>) + 584
27 swift-frontend           0x0000000103e8b7dc swift::TypeChecker::typeCheckExpression(swift::constraints::SyntacticElementTarget&, swift::OptionSet<swift::TypeCheckExprFlags, unsigned int>) + 436
28 swift-frontend           0x0000000103e8cdd0 swift::TypeChecker::typeCheckBinding(swift::Pattern*&, swift::Expr*&, swift::DeclContext*, swift::Type, swift::PatternBindingDecl*, unsigned int, swift::OptionSet<swift::TypeCheckExprFlags, unsigned int>) + 172
29 swift-frontend           0x0000000103f972e0 swift::TypeChecker::typeCheckStmtConditionElement(swift::StmtConditionElement&, bool&, swift::DeclContext*) + 236
30 swift-frontend           0x0000000103fa025c typeCheckConditionForStatement(swift::LabeledConditionalStmt*, swift::DeclContext*) + 88
31 swift-frontend           0x0000000103f9e558 swift::ASTVisitor<(anonymous namespace)::StmtChecker, void, swift::Stmt*, void, void, void, void>::visit(swift::Stmt*) + 1116
32 swift-frontend           0x0000000103f9dfc4 bool (anonymous namespace)::StmtChecker::typeCheckStmt<swift::Stmt>(swift::Stmt*&) + 316
33 swift-frontend           0x0000000103f99008 (anonymous namespace)::StmtChecker::typeCheckASTNode(swift::ASTNode&) + 104
34 swift-frontend           0x0000000103f9e224 swift::ASTVisitor<(anonymous namespace)::StmtChecker, void, swift::Stmt*, void, void, void, void>::visit(swift::Stmt*) + 296
35 swift-frontend           0x0000000103f9ba6c bool (anonymous namespace)::StmtChecker::typeCheckStmt<swift::BraceStmt>(swift::BraceStmt*&) + 316
36 swift-frontend           0x0000000103f9bc40 swift::TypeChecker::typeCheckTopLevelCodeDecl(swift::TopLevelCodeDecl*) + 140
37 swift-frontend           0x0000000103fe8498 swift::TypeCheckSourceFileRequest::evaluate(swift::Evaluator&, swift::SourceFile*) const + 472
38 swift-frontend           0x0000000103fee988 llvm::Expected<swift::TypeCheckSourceFileRequest::OutputType> swift::Evaluator::getResultUncached<swift::TypeCheckSourceFileRequest>(swift::TypeCheckSourceFileRequest const&) + 664
39 swift-frontend           0x0000000103fe8240 swift::TypeCheckSourceFileRequest::OutputType swift::evaluateOrDefault<swift::TypeCheckSourceFileRequest>(swift::Evaluator&, swift::TypeCheckSourceFileRequest, swift::TypeCheckSourceFileRequest::OutputType) + 228
40 swift-frontend           0x0000000102d788dc swift::CompilerInstance::performSema() + 332
41 swift-frontend           0x0000000102b7d9c4 withSemanticAnalysis(swift::CompilerInstance&, swift::FrontendObserver*, llvm::function_ref<bool (swift::CompilerInstance&)>, bool) + 40
42 swift-frontend           0x0000000102b6df0c performCompile(swift::CompilerInstance&, int&, swift::FrontendObserver*) + 3616
43 swift-frontend           0x0000000102b6b6d0 swift::performFrontend(llvm::ArrayRef<char const*>, char const*, void*, swift::FrontendObserver*) + 4968
44 swift-frontend           0x0000000102afae8c swift::mainEntry(int, char const**) + 2612
45 dyld                     0x000000019bade0e0 start + 2360
zsh: segmentation fault  swift toto.swift
```

### Expected behavior

`swiftc` should compile this code (or fail without crashing; I’m not sure it is actually possible to solve the generics, but it should be),

### Environment

swift-driver version: 1.90.11.1 Apple Swift version 5.10 (swiftlang-5.10.0.13 clang-1500.3.9.4)
Target: arm64-apple-macosx14.0

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by AnthonyLatsis**
The only case a singleton metatype transforms into an existential metatype is when the instance type transforms into an existential from a non-existential.

Resolves #72941.

___

NB: Type erasure correctness is not enough to fix crashes in cases like the one in the linked issue — `A?.Type` to `B?.Type` where `A <: B` — because the compiler does not support these correct conversions to begin with. With nothing stopping them from surviving the constraint solver, they either trip an assertion somewhere down the line or sometimes even compile. Handle these cases gracefully by requiring the constraint system to certify conversions associated with existential member accesses.

### CODE REVIEW

File: test/Constraints/issue-67337.swift
Comment (AnthonyLatsis) [inline test/Constraints/issue-67337.swift:1]: Incorporated this test into `test/decl/protocol/existential_member_accesses_self_assoctype.swift` [here](https://github.com/apple/swift/pull/73103/files#diff-4a16d27ba7678f839a4b9f96eaf035c1d4fb25d585d14ff371aa2baf87eed3b8R82) .

**Review by slavapestov — APPROVED** [review]

**Comment by AnthonyLatsis** [comment]
@swift-ci please smoke test macOS

**Comment by AnthonyLatsis** [comment]
@swift-ci please smoke test Linux

**Comment by AnthonyLatsis** [comment]
@swift-ci please test source compatibility release

**Comment by AnthonyLatsis** [comment]
@swift-ci please smoke test Windows

**Comment by AnthonyLatsis** [comment]
@swift-ci please smoke test macOS

**Pull request merged**